WIP microsites from database, branding enforcement and microsite settings object

felipemontoya opened the pull request:
In it's current form this PR is only for review purposes. 
For those looking at it, it does 3 things:

- Adds the ability to define the microsites as a record in the database instead of the current definition at the settings file.
- Adds a middleware to prevent courses defined in a ORG from a microsite to be displayed in other microsite or a non-microsited view
- Adds a microsite aware settings object to be used as a helper when dealing with microsite support



Notes:
- I can break this into 3 PRs if it is necessary
- Since the starting commit is the last commit from the birch branch, there are 3 commits by the edx team that probably don't belong here.

edx-webhook commented:
Thanks for the pull request, @felipemontoya! I've created [OSPR-565](https://openedx.atlassian.net/browse/OSPR-565) to keep track of it in JIRA. JIRA is a place for product owners to prioritize feature reviews by the engineering development teams. 

Feel free to add as much of the following information to the ticket:
- supporting documentation
- edx-code email threads
- timeline information ('this must be merged by XX date', and why that is)
- partner information ('this is a course on edx.org')
- any other information that can help Product understand the context for the PR

All technical communication about the code itself will still be done via the Github pull request interface. As a reminder, [our process documentation is here](http://edx-developer-guide.readthedocs.org/en/latest/process/overview.html).

felipemontoya commented:
Hi @chrisndodge, I cherry picked the changes here to make them easier to review/comment on. 
Also pinging @nedbat and @e0d in case you want to look at this.

sarina commented:
@felipemontoya -- unfortunately, things aren't going to work if you have a PR based off of Birch. You're going to have to create a development version that's based on the current version of master.

The CI build simply won't run correctly anymore off of a Birch-based branch, since we make many changes to our test infrastructure. Plus, you have a fair bit of merge conflicts. So, please figure out the best way to do that.

Have you talked to people at edX about this work? If so, please include any links to mailing list discussions or any other supporting information that you have.

sarina commented:
You will see I cancelled your CI build - that's because it's going to run for about two hours before hanging and giving up. It needs to be rebased on master to run.

chrisndodge commented:
@sarina, Felipe reached out to Ned last week while I was out on vacation. The other day I said that I'd take a look at it when we had an initial PR ready to make commenting easier.

felipemontoya commented:
@sarina Thanks, I actually thought of cancelling it myself but didn't know how or if it was at all posible.

sarina commented:
@felipemontoya yeah unfortunately only we can cancel the build. Happy to help with any other CI questions that may come up. I'm out of the office until Monday but I'll catch up with this when I'm back.

chrisndodge commented on common/djangoapps/microsite_configuration/__init__.py:
```diff
@@ -1 +1,19 @@
+    def __getattr__(self, name):
+        return microsite.get_value(name, base_settings.__getattr__(name))
+
+settings = MicrositeAwareSettings()
```
I like this very much. We struggle where developers don't set up their calls to use microsite.get_value(), this approach at leasts allows the python to remain consistent: settings.MY_KEY. We then just need to make sure the imports are updated appropriately.

chrisndodge replied:
Thinking a bit more, this approach does expose a mistake I've made in the beginning of Microsites. One of the most common config parameters to override is PLATFORM_NAME, but as a config override I did not preserve the capitalization, so it appears in microsite configuration as "platform_name". I think think gap is significant as your __getattr__ wouldn't work in that case, because the attribute/key names don't match exactly.

Let me think about the significance of this. It might be best to bite the bullet and rectify this, but I'd like to think through what that might mean from a practical point of view.

felipemontoya replied:
We where recently bit by this. We created an override of all the templates using the capitalized form(devstack had the microsite config capitalized) and later on, it failed on stage. From what I recall of this day, there where about ~120 occurrences of this. I would support rectifying this now.

felipemontoya replied:
The one thing that I feel is missing with this microsite aware settings object is the fact that it would allow a microsite to override ANY setting. That could open too much space for microsite to break things. I could add a filter to this in the form of a list of allowed_overrides. If a given name is not in this list then it would directly take the global setting for it.

chrisndodge replied:
Recify - which would entail a lot of touch points - or add a "key translation" (e.g. "PLATFORM_NAME" = "platform_name") dictionary here. It seems less elegant, but might make for a better initial migration point.

chrisndodge commented on common/djangoapps/microsite_configuration/middleware.py:
```diff
@@ -83,3 +87,54 @@ def _set_cookie_wrapper(key, value='', max_age=None, expires=None, path='/', dom
+    """
+    Middleware class that prevents a course defined in a branded ORG trough a microsite, to be displayed
+    on a different microsite with a different branding.
+    """
```
I'm not sure how this will work when Studio goes to the LMS to preview draft content, because I think we use a hostname of 'preview.edx.org' or something like that. I'll have to investigate a bit more closely.

chrisndodge replied:
Just to follow up on this comment, I would presume your team would have encountered this already (regarding support for a 'preview' from Studio), no?

felipemontoya replied:
Since for us the 'preview' button from studio points the instance without branding, and our customers have some heavy branding, we discourage the use of said button. So the request are indeed blocked, but it is not a big issue for us now.

What could be done here would be to either leave the use of this filter as a per microsite configuration, or to add a whitelist of domains who can render any course.

chrisndodge commented on requirements/edx/base.txt:
```diff
@@ -56,7 +56,7 @@ nose==1.3.3
 paramiko==1.9.0
 path.py==3.0.1
-Pillow==2.6.1
+Pillow==2.7.0
```
Is this version change necessary?

felipemontoya replied:
This comes from the differences on the birch branch vs master. It should be gone when I rebase this to master.

chrisndodge commented:
@felipemontoya I hope you don't mind if I tag @natea on this PR as well. As I mentioned earlier, his team has been working with Microsites a lot.

chrisndodge commented on common/djangoapps/microsite_configuration/microsite.py:
```diff
@@ -98,13 +121,23 @@ def get_value_for_org(org, val_name, default=None):
+            if org_filter:
+                org_filter_set.add(org_filter)
+
+    # Get the orgs in the settings file
```
I would prefer setting up an interface layer here rather than doing conditional logic. Seems like we need a MicrositeConfigurationBackend abstract interface with two implementations of the interface: MicrositeDjangoSettingsConfigurationBackend and MicrositeDatabaseConfigurationBackend.

What do you think about that? I don't think the interface would need a lot of methods.

felipemontoya replied:
I like that. That would mean there is a django.setting that selects the microsite backend and these get_all_orgs method, and posible get_value_for_org only call the backend method, right?

I think these methods would be required:
- has_configuration_set
- get_template_path (posibly)
- get_value_for_org
- get_all_orgs
- _set_current_microsite
- set_by_domain

Then we could remove set_from_db_by_domain, and _set_current_microsite_from_obj

Do you think that would make this, the sort of PR that changes too much and then it is too dificult to review?

sarina replied:
@felipemontoya you could make the changes in the PR and if it's too big, maybe think about the best way to separate them into smaller PRs.

Frankly I'd recommend rebasing on master as soon as you can. I'm worried that if you keep implementing more things, there may be conflicts (there definitely are already a few) that will take a lot of development work to resolve.

felipemontoya commented:
@chrisndodge Last week was too busy for me here, so I could not look much into this PR, I just rebased it to master and it worked locally, so I pushed it for updateness sake, and will be addressing your feedback this week.

chrisndodge commented:
Hi @felipemontoya just checking back in to pick up this conversation again. What are your current thoughts about this PR?

One thing I was thinking of is to separate some things out into separate PR's. For example, the changes to expose a call-compatible interface to Django's settings module (MicrositeAwareSettings) could be useful and a very isolated additive change (i.e. current code can continue to work as is).

We're having some challenges with non-microsite aware configuration lookup patterns, and I think simply having people use a different 'import' statement but the same normal settings lookup syntax would probably help make things a bit less error prone.

FYI, @benpatterson @cahrens @andy-armstrong

felipemontoya commented:
Hello @chrisndodge, I think thats something I could prioritize. I have been a bit slow keeping up with this PR, and the required development is required only for the microsites in the database part . But the MicrositesAwareSettings object is ready to go. 
I'll split this PR then into 2 then. 

For now I'll leave the branding filter out, since it could break the preview and to properly fix this, I need to go into detail in studio, which would make it too slow to get done. If someone reads this and find a particular interest in the branding filter, let me know and I'll create a PR for that.

chrisndodge commented:
@andy-armstrong @cahrens what do you think of the above. Basically the proposal is to allow for coders to do (roughly):

```
from microsites_config.settings import MicrositesAwareSettings as settings
:
:
settings.{config_name}
:
```

rather than having to always remember to do microsites.get_value(...)

Seems easier to police using the right import than the 1000's of instances where we call settings.{foo}

cahrens commented:
@chrisndodge So the idea then is that we ALWAYS use microsites_config.settings instead of "regular" settings (regardless of whether or not a setting is something that microsites would wish to override)? If so, then :+1:

Would the PR include change all of our existing usages of settings to microsites_config.settings?

sarina commented:
@felipemontoya can this WIP pull request be closed now that #8971 is opened?

felipemontoya commented:
Hi @sarina. I would say yes to that. There is a last thing here, that is not in #8971, but I think it did not catch much momentum, so the important bits are already the other 2 PRs
